### PR TITLE
Base for billing API

### PR DESCRIPTION
**Pull request by rockambole86**

### CODE REVIEW

File: Billing.API.Test/InvoiceApiTest.cs [hunk 1]
```diff
@@ -1,22 +1,22 @@
...
-using Microsoft.AspNetCore.Mvc.Testing;
 using System;
 using System.Collections.Generic;
 using System.Net;
 using System.Net.Http;
 using System.Net.Http.Headers;
 using System.Threading.Tasks;
+using AutoFixture.Xunit2;
+using Flurl.Http.Testing;
+using Microsoft.AspNetCore.Mvc.Testing;
```
Comment (andresmoschini) [inline Billing.API.Test/InvoiceApiTest.cs:9]: Creo que el standard que se está usando es ordenar alfabeticamente (no system primero)

Comment (rockambole86) [reply Billing.API.Test/InvoiceApiTest.cs:9]: ahhh, joya, ya lo arreglo, lo tenía seteado automático

File: Billing.API.Test/InvoiceApiTest.cs [hunk 2]
```diff
@@ -27,37 +27,9 @@ public void Dispose()
...
-            ParticipacionesAccionarias = true,
-            PersonaFisica = false,
-            RazonSocial = "RZS C.S. SA",
-            StatCode = 0
-        };
-
         [Theory]
         [AutoData]
-        public async Task GET_random_URL_should_return_404_NotFound(string url)
+        public async Task GetInvoices_ReturnsNotFound(string url)
```
Comment (andresmoschini) [inline Billing.API.Test/InvoiceApiTest.cs:32]: What do you think about explaining what conditions (_when_) it returns not found?

> Unit test method names must be descriptive about _what is being tested_, under _what conditions_, and _what the expectations are_. Pascal casing and underscores can be used to improve readability.

File: Billing.API/Billing.API.csproj [hunk 3]
```diff
@@ -7,14 +7,15 @@
...
-    <FileVersion>2020.7.27.2</FileVersion>
+    <AssemblyVersion>2020.7.30.0</AssemblyVersion>
+    <FileVersion>2020.7.30.0</FileVersion>
   </PropertyGroup>
 
   <ItemGroup>
     <PackageReference Include="Flurl.Http" Version="2.4.2" />
     <PackageReference Include="Microsoft.AspNetCore.Authentication.JwtBearer" Version="3.1.6" />
     <PackageReference Include="Tavis.UriTemplates" Version="1.1.1" />
+    <PackageReference Include="Z.ExtensionMethods" Version="2.1.1" />
```
Comment (andresmoschini) [inline Billing.API/Billing.API.csproj:18]: What is this used for?

Comment (rockambole86) [reply Billing.API/Billing.API.csproj:18]: Ahhhh... that's a library I include in almost any project, it provides useful extension methods for almost any base types... Making them Fluent style.
For example:

```csharp
string.IsNullOrEmpty(variable) -> variable.IsNullOrEmpty() 

variable.In(value1, value2)
```

Honestly, I love it. And it's from the creators of Dapper

File: Billing.API/Controllers/InvoiceController.cs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        {
+            _logger = logger;
+            _invoiceService = invoiceService;
+        }
+
+        [HttpGet("/invoices/{clientId}")]
+        public async Task<IActionResult> GetInvoices([FromRoute] string clientId)
+        {
+            if (clientId.Length != 15)
+                return BadRequest();
```
Comment (andresmoschini) [inline Billing.API/Controllers/InvoiceController.cs:29]: Could you validate it declaratively with a _Validation Attribute_ or a _Route Contraint_?

I am not sure, but something like these should work:

```
[HttpGet("/invoices/{clientId:long:length(15)}")]
public async Task<IActionResult> GetInvoices([FromRoute] string clientId)
```

or 

```
[HttpGet("/invoices/{clientId}")]
public async Task<IActionResult> GetInvoices([FromRoute][Length(15)] string clientId)
```

File: Billing.API/CuitNumber.cs [hunk 5]
```diff
@@ -14,31 +14,36 @@ public class CuitNumber : IValidatableObject
         // TODO: rename as OriginalValue
         // Ugly patch, this value should come from de validated parameter, ie: /taxinfo/by-cuit/{cuit}
         // it wil be fixed after moving to a custom model binder
-        public string? cuit { get; set; }
-        public string SimplifiedValue => cuit?.Replace("-", "") ?? string.Empty;
+        public string? CUIT { get; set; }
+
+        public string SimplifiedValue => CUIT?.ReplaceByEmpty("-") ?? string.Empty;
```
Comment (andresmoschini) [inline Billing.API/CuitNumber.cs:19]: I think that you can remove this file.

File: Billing.API/DopplerSecurity/DopplerSecurityServiceCollectionExtensions.cs [hunk 6]
```diff
@@ -1,10 +1,10 @@
-using Billing.API.DopplerSecurity;
 using Microsoft.AspNetCore.Authentication.JwtBearer;
 using Microsoft.AspNetCore.Authorization;
+using Microsoft.Extensions.DependencyInjection;
 using Microsoft.Extensions.Options;
 using Microsoft.IdentityModel.Tokens;
 
-namespace Microsoft.Extensions.DependencyInjection
+namespace Billing.API.DopplerSecurity
```
Comment (andresmoschini) [inline Billing.API/DopplerSecurity/DopplerSecurityServiceCollectionExtensions.cs:7]: You can change it if you want, but the previous name was fine: https://github.com/FromDoppler/cuit-service/pull/7#discussion_r454048187

File: Billing.API/CuitNumber.cs (hunk 5 above)
Comment (rockambole86) [reply Billing.API/CuitNumber.cs:19]: The whole class, right?

File: Billing.API/Models/TaxInfo.cs [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+        public string? Message                    { get; set; }
+        public bool    Monotributo                { get; set; }
+        public string? MonotributoActividad       { get; set; }
+        public string? MonotributoCategoria       { get; set; }
+        public string? Nombre                     { get; set; }
+        public string? PadronData                 { get; set; }
+        public bool    ParticipacionesAccionarias { get; set; }
+        public bool    PersonaFisica              { get; set; }
+        public string? RazonSocial                { get; set; }
+        public int     StatCode                   { get; set; }
```
Comment (andresmoschini) [inline Billing.API/Models/TaxInfo.cs:28]: If you really like it, you can keep this format, but in general, I prefer a non-tabulated format, it is better to resolve conflicts and to minify the diff.

But, maybe you can remove this file, it seems to be unrelated to the scope of your service.

File: Billing.API/Startup.cs [hunk 8]
```diff
@@ -18,7 +20,7 @@ public Startup(IConfiguration configuration)
         public void ConfigureServices(IServiceCollection services)
         {
             services.AddDopplerSecurity();
-            services.AddTaxInfoProvider();
+            services.AddInvoice();
```
Comment (andresmoschini) [inline Billing.API/Startup.cs:23]: Very minor, and opinable, of course:

```suggestion
            services.AddInvoiceService();
```

```suggestion
            services.AddInvoicesService();
```

```suggestion
            services.AddInvoicesProvider();
```

File: Billing.API/appsettings.json [hunk 9]
```diff
@@ -8,8 +8,8 @@
   },
   "AllowedHosts": "*",
   "PublicKeysFolder": "public-keys",
-  "TaxInfoProvider": {
-    "UriTemplate": "http://{host}:33333/api/TaxInfo?ID={cuit}+CardData=Y",
+  "Invoice": {
+    "UriTemplate": "http://{host}:33333/api/Invoice?clientId={clientId}",
```
Comment (andresmoschini) [inline Billing.API/appsettings.json:12]: I think that you can remove this entry, you will probably query the backend in a totally different way.

File: Billing.API/Billing.API.csproj (hunk 3 above)
Comment (andresmoschini) [reply Billing.API/Billing.API.csproj:18]: Mmm, ok, only be sure to verify that it does not include potentially conflictive dependencies.

File: Billing.API/CuitNumber.cs (hunk 5 above)
Comment (andresmoschini) [reply Billing.API/CuitNumber.cs:19]: Yes, if I am not wrong, you will not need it.

File: Billing.API/Models/TaxInfo.cs (hunk 7 above)
Comment (rockambole86) [reply Billing.API/Models/TaxInfo.cs:28]: Removed

File: Billing.API/Startup.cs (hunk 8 above)
Comment (rockambole86) [reply Billing.API/Startup.cs:23]: Changed...
Three times? lols

**Comment by nreal** [comment]
Good job 👍

File: Billing.API/Controllers/InvoiceController.cs [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+                    zipStream.CloseEntry();
+
+                    zipStream.IsStreamOwner = false;
+                }
+            }
+
+            ms.Position = 0;
+
+            return File(ms, "application/octet-stream", $"Ïnvoices_{clientId}.zip");
```
Comment (andresmoschini) [inline Billing.API/Controllers/InvoiceController.cs:60]: Why are you returning a zip file in place of a JSON list?

File: Billing.API/appsettings.json [hunk 11]
```diff
@@ -8,10 +8,11 @@
...
   "AllowedHosts": "*",
   "PublicKeysFolder": "public-keys",
-  "TaxInfoProvider": {
-    "UriTemplate": "http://{host}:33333/api/TaxInfo?ID={cuit}+CardData=Y",
+  "Invoice": {
     "Host": "REPLACE_FOR_REAL_HOST",
     "UserName": "REPLACE_FOR_REAL_USERNAME",
-    "Password": "REPLACE_FOR_REAL_PASSWORD"
+    "Password": "REPLACE_FOR_REAL_PASSWORD",
+    "DbConnectionString": "Server=172.25.16.86:30015;UserName=DOPPLER;Password=uz2_NDgV4h#=Hx6;",
```
Comment (andresmoschini) [inline Billing.API/appsettings.json:15]: Take into account that this code is totally public, is it not a secret, right?

Comment (andresmoschini) [reply Billing.API/appsettings.json:15]: In case that it is a secret, we will need to change the password.

Comment (andresmoschini) [inline Billing.API/appsettings.json:15]: If we need to store this information in encrypted secrets, it would be nice to split this information in more than a field and only encrypt the secret ones.

File: Billing.API/Services/Invoice/IInvoiceService.cs [hunk 12]
```diff
@@ -0,0 +1,12 @@
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace Billing.API.Services.Invoice
+{
+    public interface IInvoiceService
+    {
+        Task<IEnumerable<string>> GetInvoices(string clientId);
```
Comment (andresmoschini) [inline Billing.API/Services/Invoice/IInvoiceService.cs:8]: If I am not wrong, the responsibility of this method is only returning a list of invoices with some information, the required data to generate a result like this:

```
GET /accounts/{doppler/relay}/{id}/invoices?from=2020-01-01T00:00:00Z&to=2020-07-01T00:00:00Z
Authorization: Bearer {{token-jwt}}
​```

```json
{
    "invoices": [
        {
            "product": "doppler",
            "accountId": 22235,
            "date": "2020-01-01T00:00:00Z",
            "currency": "ARS",
            "amount": 128,
            "_links": [
                {
                    "rel": "file",
                    "path": "/accounts/doppler/22235/invoices/00000121544.pdf"
                }
            ]
        },
    ],
    "_links": [
        {
            "rel": "next",
            "path": "/accounts/doppler/22235/invoices?from=2020-02-01T00:00:00Z&to=2020-07-01T00:00:00Z"
        }
    ]
}
```

Comment (andresmoschini) [reply Billing.API/Services/Invoice/IInvoiceService.cs:8]: `/accounts/doppler/22235/invoices/00000121544.pdf` is the public route, it is not necessary couple it to the backend path.

**Comment by andresmoschini** [comment]
image

Did you research about this error?
